### ISSUE TITLE
Support for immutable variables 

### ISSUE THREAD

**Issue by tristanhimmelman** [Hearst-DD/ObjectMapper#383]
I'm creating this issue as a place to discuss how we might support immutable variables.

ObjectMapper already has some rudimentary support of immutable types. There is some documentation hidden within the README which you can reference.

**Comment by Sega-Zero** [comment]
#384 tries to solve this, but it breaks the ability to add `Mappable` conformance via extension. I solved the immutable mapping problem with a little bit hacky, but pretty much working code:
```swift
//somewhere else in code
struct SampleStruct {
    let avatar: String
    let name: String
}

//another swift file that gathers all models that needs to be Mappable
extension SampleStruct: ImmutableMappable {
    static func map(map: Map) -> SampleStruct? {
        var mapAvatar = ""; mapAvatar <- map["avatar"]
        var mapName = ""; mapName <- map["name"]

        return SampleStruct(avatar: mapAvatar, name: mapName)
    }
}

// MARK: private boilerplate code

protocol ImmutableMappable {
    static func map(map: Map) -> Self?
}

private struct MappableStub: Mappable {
    var map: Map?

    init?(_ map: Map) {
    }

    // Mappable
    mutating func mapping(map: Map) {
        self.map = map
    }
}

extension ImmutableMappable {
    static func mappedInstance(json: String) -> Self? {
        guard let stub: MappableStub? = Mapper().map(json), let map = stub?.map else { return nil }
        return self.map(map)
    }
}
```

Now all I need is call `let sample = SampleStruct.mappedInstance(json)`

**Comment by mpdomas** [comment]
We use a more straightforward method.
First we introduce an extension for `Map` that adds `typedValue()` function:
```swift
import ObjectMapper

extension Mappable {
    mutating func mapping(map: Map) {}
}
extension Map {
    func typedValue<T>()          ->  T?  { var v: T?; v <- self; return v }
    func typedValue<T>()          -> [T]? { var v: T?; v <- self; return v }
    func typedValue<T:Mappable>() ->  T?  { var v: T?; v <- self; return v }
    func typedValue<T:Mappable>() -> [T]? { var v: T?; v <- self; return v }
}
```
Then our imutable variable can be initialized like so:
```swift
import ObjectMapper

struct Human: Mappable {
    init?(_ map: Map) {
        guard let
            _name: String = map["name"].typedValue(),
            _age:  Int =    map["age"] .typedValue()
            else { return nil }
        
        name = _name
        age  = _age
    }
    
    let name: String
    let age:  Int
}
```
The only problem with this approach is when you need to support `toJSON()` then we add a `mapping` function:
```swift
// Variables have to be var, not let, or even more fiddling is required
mutating func mapping(map: Map) {
    guard map.mappingType == .ToJSON else { return }
    
    name <- map["name"]
    age  <- map["age"]
}
```

**Comment by mwawrusch** [comment]
We just do a

```
private(set) var xx....
```

**Comment by tristanhimmelman** [comment]
Support for immutable variables has just been merged into master 👍  #592 . Release 2.1.0 has the latest. Thanks again @devxoul
